Qute template extensions throw ambiguous exception on property not found

### Describe the bug

"{time:format(input.birthday, 'uuuu')}" throw IllegalStateException without any meaningful message in case the "birthday" property does not exist.

### Expected behavior

should convey meaningful message similar to what {input.birthday} will throw:

`io.quarkus.qute.TemplateException: Property "birthday" not found on the base object "java.util.ImmutableCollections$Map1" in expression {input.birthday} in template 1 on line 1

	at io.quarkus.qute.EvaluatorImpl.propertyNotFound(EvaluatorImpl.java:208)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:169)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolve$4(EvaluatorImpl.java:179)
	at io.quarkus.qute.CompletedStage.thenCompose(CompletedStage.java:228)
	at io.quarkus.qute.EvaluatorImpl.resolve(EvaluatorImpl.java:176)
	at io.quarkus.qute.EvaluatorImpl.resolveReference(EvaluatorImpl.java:113)
	at io.quarkus.qute.EvaluatorImpl.lambda$resolveReference$2(EvaluatorImpl.java:117)
	at io.quarkus.qute.CompletedStage.thenCompose(CompletedStage.java:228)
	at io.quarkus.qute.EvaluatorImpl.resolveReference(EvaluatorImpl.java:117)
	at io.quarkus.qute.EvaluatorImpl.evaluate(EvaluatorImpl.java:81)
	at io.quarkus.qute.ResolutionContextImpl.evaluate(ResolutionContextImpl.java:28)
	at io.quarkus.qute.ExpressionNode.resolve(ExpressionNode.java:25)
	at io.quarkus.qute.SectionNode$SectionResolutionContextImpl.execute(SectionNode.java:149)
	at io.quarkus.qute.SectionHelper$SectionResolutionContext.execute(SectionHelper.java:37)
	at io.quarkus.qute.Parser$1.resolve(Parser.java:1054)
	at io.quarkus.qute.SectionNode.resolve(SectionNode.java:34)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderData(TemplateImpl.java:104)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.renderAsync(TemplateImpl.java:90)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.render(TemplateImpl.java:55)
	at TimeTemplateExtensionsTest.test(TimeTemplateExtensionsTest.java:33)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.test.junit.QuarkusTestExtension.runExtensionMethod(QuarkusTestExtension.java:956)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:776)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)`

### Actual behavior

`Caused by: java.lang.IllegalStateException: java.util.concurrent.TimeoutException
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.render(TemplateImpl.java:60)
	at TimeTemplateExtensionsTest.lambda$throwAmbiguousExceptionOnPropertyNotFound$0(TimeTemplateExtensionsTest.java:23)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:55)
	... 74 more
Caused by: java.util.concurrent.TimeoutException
	at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1886)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2021)
	at io.quarkus.qute.TemplateImpl$TemplateInstanceImpl.render(TemplateImpl.java:55)
	... 76 more`

### How to Reproduce?

Please look at the test class TimeTemplateExtensionsTest
[test-quarkus.zip](https://github.com/quarkusio/quarkus/files/7878677/test-quarkus.zip)


### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

openjdk 11

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

Quarkus 2.6.2.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/docs/src/main/asciidoc/qute-reference.adoc
+++ b/docs/src/main/asciidoc/qute-reference.adoc
@@ -1572,7 +1572,7 @@ TIP: A list element can be accessed directly: `{list.10}` or `{list[10]}`.
 * `mod`: Modulo operation
 ** `{#if counter.mod(5) == 0}`
 
-==== Strings
+===== Strings
 
 * `fmt` or `format`: format the string instance via `java.lang.String.format()`
 ** `{myStr.fmt("arg1","arg2")}`
@@ -1616,7 +1616,7 @@ TIP: A list element can be accessed directly: `{list.10}` or `{list[10]}`.
 ** `{time:format(myDate,'d MMM uuuu', myLocale)}`
 
 * `time:format(dateTime,pattern,locale,timeZone)`: Formats temporal objects from the `java.time` package, `java.util.Date`, `java.util.Calendar` and `java.lang.Number`
-** `{time:format(myDate,'d MMM uuuu',myTimeZoneId)}`
+** `{time:format(myDate,'d MMM uuuu',myLocale,myTimeZoneId)}`
 
 [[template_data]]
 === @TemplateData

--- a/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/extensions/TimeTemplateExtensionsTest.java
+++ b/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/extensions/TimeTemplateExtensionsTest.java
@@ -1,12 +1,15 @@
 package io.quarkus.qute.deployment.extensions;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
 
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.util.Calendar;
 import java.util.Date;
 import java.util.Locale;
+import java.util.Map;
 
 import javax.inject.Inject;
 
@@ -14,7 +17,9 @@ import org.jboss.shrinkwrap.api.asset.StringAsset;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
+import io.quarkus.qute.Engine;
 import io.quarkus.qute.Template;
+import io.quarkus.qute.TemplateException;
 import io.quarkus.test.QuarkusUnitTest;
 
 public class TimeTemplateExtensionsTest {
@@ -29,6 +34,9 @@ public class TimeTemplateExtensionsTest {
     @Inject
     Template foo;
 
+    @Inject
+    Engine engine;
+
     @Test
     public void testFormat() {
         Calendar nowCal = Calendar.getInstance();
@@ -42,6 +50,17 @@ public class TimeTemplateExtensionsTest {
                 .data("nowDate", nowDate)
                 .data("nowCalendar", nowCal)
                 .data("myLocale", Locale.ENGLISH).render());
+    }
+
+    @Test
+    public void testInvalidParameter() {
+        try {
+            // input.birthday cannot be resolved
+            engine.parse("{time:format(input.birthday, 'uuuu')}").data("input", Map.of("name", "Quarkus Qute")).render();
+            fail();
+        } catch (TemplateException expected) {
+            assertTrue(expected.getMessage().startsWith("Property \"birthday\" not found on the base object"));
+        }
     }
 
 }

--- a/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/extensions/TimeTemplateExtensions.java
+++ b/extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/extensions/TimeTemplateExtensions.java
@@ -23,6 +23,10 @@ public class TimeTemplateExtensions {
 
     private static final Map<Key, DateTimeFormatter> FORMATTER_CACHE = new ConcurrentHashMap<>();
 
+    public static void clearCache() {
+        FORMATTER_CACHE.clear();
+    }
+
     static String format(TemporalAccessor temporal, String pattern) {
         return FORMATTER_CACHE.computeIfAbsent(new Key(pattern, null, null), TimeTemplateExtensions::formatterForKey)
                 .format(temporal);
@@ -89,21 +93,23 @@ public class TimeTemplateExtensions {
         private final String pattern;
         private final Locale locale;
         private final ZoneId timeZone;
+        private final int hashCode;
 
         public Key(String pattern, Locale locale, ZoneId timeZone) {
             this.pattern = pattern;
             this.locale = locale;
             this.timeZone = timeZone;
-        }
-
-        @Override
-        public int hashCode() {
             final int prime = 31;
             int result = 1;
             result = prime * result + ((locale == null) ? 0 : locale.hashCode());
             result = prime * result + ((pattern == null) ? 0 : pattern.hashCode());
             result = prime * result + ((timeZone == null) ? 0 : timeZone.hashCode());
-            return result;
+            this.hashCode = result;
+        }
+
+        @Override
+        public int hashCode() {
+            return hashCode;
         }
 
         @Override

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/CompletedStage.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/CompletedStage.java
@@ -38,6 +38,10 @@ public final class CompletedStage<T> implements CompletionStage<T>, Supplier<T> 
         this.exception = exception;
     }
 
+    public boolean isFailure() {
+        return exception != null;
+    }
+
     public T get() {
         if (exception != null) {
             // Throw an exception if completed exceptionally

--- a/independent-projects/qute/core/src/main/java/io/quarkus/qute/EvaluatedParams.java
+++ b/independent-projects/qute/core/src/main/java/io/quarkus/qute/EvaluatedParams.java
@@ -9,7 +9,6 @@ import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
 import java.util.function.Supplier;
 
-@SuppressWarnings("rawtypes")
 public final class EvaluatedParams {
 
     static final EvaluatedParams EMPTY = new EvaluatedParams(CompletedStage.VOID, new Supplier<?>[0]);
@@ -29,12 +28,17 @@ public final class EvaluatedParams {
         Supplier<?>[] allResults = new Supplier[params.size()];
         List<CompletableFuture<?>> asyncResults = null;
         int i = 0;
+        CompletedStage<?> failure = null;
         Iterator<Expression> it = params.iterator();
         while (it.hasNext()) {
             Expression expression = it.next();
             CompletionStage<?> result = context.evaluate(expression);
             if (result instanceof CompletedStage) {
-                allResults[i++] = (CompletedStage<?>) result;
+                CompletedStage<?> completed = (CompletedStage<?>) result;
+                allResults[i++] = completed;
+                if (completed.isFailure()) {
+                    failure = completed;
+                }
                 // No async computation needed
                 continue;
             } else {
@@ -48,7 +52,7 @@ public final class EvaluatedParams {
         }
         CompletionStage<?> cs;
         if (asyncResults == null) {
-            cs = CompletedStage.VOID;
+            cs = failure != null ? failure : CompletedStage.VOID;
         } else if (asyncResults.size() == 1) {
             cs = asyncResults.get(0);
         } else {
@@ -71,18 +75,23 @@ public final class EvaluatedParams {
             return EMPTY;
         }
         Supplier<?>[] allResults = new Supplier[params.size()];
-        List<CompletableFuture<Object>> asyncResults = null;
+        List<CompletableFuture<?>> asyncResults = null;
 
         int i = 0;
+        CompletedStage<?> failure = null;
         Iterator<Expression> it = params.subList(1, params.size()).iterator();
         while (it.hasNext()) {
-            CompletionStage<Object> result = context.evaluate(it.next());
+            CompletionStage<?> result = context.evaluate(it.next());
             if (result instanceof CompletedStage) {
-                allResults[i++] = (CompletedStage<Object>) result;
+                CompletedStage<?> completed = (CompletedStage<?>) result;
+                allResults[i++] = completed;
+                if (completed.isFailure()) {
+                    failure = completed;
+                }
                 // No async computation needed
                 continue;
             } else {
-                CompletableFuture<Object> fu = result.toCompletableFuture();
+                CompletableFuture<?> fu = result.toCompletableFuture();
                 if (asyncResults == null) {
                     asyncResults = new ArrayList<>();
                 }
@@ -92,7 +101,7 @@ public final class EvaluatedParams {
         }
         CompletionStage<?> cs;
         if (asyncResults == null) {
-            cs = CompletedStage.VOID;
+            cs = failure != null ? failure : CompletedStage.VOID;
         } else if (asyncResults.size() == 1) {
             cs = asyncResults.get(0);
         } else {
@@ -107,7 +116,7 @@ public final class EvaluatedParams {
     EvaluatedParams(CompletionStage<?> stage) {
         this.stage = stage;
         if (stage instanceof CompletedStage) {
-            this.results = new Supplier[] { (CompletedStage) stage };
+            this.results = new Supplier[] { (CompletedStage<?>) stage };
         } else {
             this.results = new Supplier[] { Futures.toSupplier(stage.toCompletableFuture()) };
         }

--- a/independent-projects/qute/core/src/test/java/io/quarkus/qute/EvaluatedParamsTest.java
+++ b/independent-projects/qute/core/src/test/java/io/quarkus/qute/EvaluatedParamsTest.java
@@ -2,8 +2,11 @@ package io.quarkus.qute;
 
 import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
 
+import java.util.List;
 import java.util.Locale;
+import java.util.concurrent.CompletionStage;
 import java.util.concurrent.ExecutionException;
 import java.util.function.Supplier;
 import org.junit.jupiter.api.Test;
@@ -32,7 +35,59 @@ public class EvaluatedParamsTest {
         assertFalse(new EvaluatedParams(null,
                 new Supplier[] { CompletedStage.of(10), CompletedStage.of("Foo") })
                         .parameterTypesMatch(false, new Class<?>[] { Integer.class, Object[].class }));
+    }
 
+    @Test
+    public void testInvalidParameter() {
+        EvaluatedParams params = EvaluatedParams.evaluate(new EvalContext() {
+
+            @Override
+            public List<Expression> getParams() {
+                Expression expr1 = ExpressionImpl.from("foo.bar");
+                Expression expr2 = ExpressionImpl.from("bar.baz");
+                return List.of(expr1, expr2);
+            }
+
+            @Override
+            public String getName() {
+                return null;
+            }
+
+            @Override
+            public Object getBase() {
+                return null;
+            }
+
+            @Override
+            public Object getAttribute(String key) {
+                return null;
+            }
+
+            @Override
+            public CompletionStage<Object> evaluate(Expression expression) {
+                if (expression.toOriginalString().equals("foo.bar")) {
+                    return CompletedStage.of("foo");
+                } else if (expression.toOriginalString().equals("bar.baz")) {
+                    return CompletedStage.failure(new IllegalArgumentException());
+                }
+                throw new IllegalStateException();
+            }
+
+            @Override
+            public CompletionStage<Object> evaluate(String expression) {
+                return null;
+            }
+        });
+        assertTrue(params.stage instanceof CompletedStage);
+        CompletedStage<?> completed = (CompletedStage<?>) params.stage;
+        assertTrue(completed.isFailure());
+        try {
+            completed.get();
+            fail();
+        } catch (TemplateException expected) {
+            Throwable cause = expected.getCause();
+            assertTrue(cause instanceof IllegalArgumentException);
+        }
     }
 
 }

